File: 백준/Silver/9934. 완전 이진 트리/완전 이진 트리.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *


input = stdin.readline


k = int(input())
n = 2**k -1

arr = list(map(int,input().split()))

tree = [[] for _ in range(k)]

k -= 1

def inOrder(child,h):
    global k 
    global tree
    
    
    root = len(child) // 2
    
    if h > k: # 높이를 벗어남
        return
    
    inOrder(child[:root],h+1)
    tree[h].append(child[root])
    inOrder(child[root+1:],h+1)


inOrder(arr,0)

for i in range(len(tree)):
    print(*tree[i])
    

File: 백준/Silver/9934. 완전 이진 트리/완전 이진 트리.swift
let k = Int(readLine()!)!

let arr = readLine()!.split{$0 == " "}.map{Int($0)!}

let n = arr.count

var tree: [[Int]] = [[Int]](repeating: [], count: k)

func inorder(_ child: [Int],_ h: Int) {
    
    if child.isEmpty {
        return
    }
    
    let root = child.count / 2
    
    
    inorder(Array(child[0..<root]), h+1)
    tree[h].append(child[root])
    inorder(Array(child[root+1..<child.count]), h+1)
    
}


inorder(arr, 0)

for i in 0..<tree.count {
    print(tree[i].map{String($0)}.joined(separator: " "))
}
